src/expr/src/expr/expr_binary_nullable.rs

```diff
@@ -62,6 +65,100 @@ macro_rules! gen_nullable_cmp_impl {
     };
 }
 
+pub struct BinaryShortCircuitExpression {
+    expr_ia1: BoxedExpression,
+    expr_ia2: BoxedExpression,
+    expr_type: Type,
+}
+
+impl std::fmt::Debug for BinaryShortCircuitExpression {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.debug_struct("BinaryShortCircuitExpression")
+            .field("expr_ia1", &self.expr_ia1)
+            .field("expr_ia2", &self.expr_ia2)
+            .field("expr_type", &self.expr_type)
+            .finish()
+    }
+}
+
+impl Expression for BinaryShortCircuitExpression {
+    fn return_type(&self) -> DataType {
+        DataType::Boolean
+    }
+
+    fn eval(&self, input: &DataChunk) -> Result<ArrayRef> {
+        let init_vis = input.vis();
+        let mut input = input.clone();
+        let len = input.capacity();
+        let mut children_array = Vec::with_capacity(2);
+        for child in [&self.expr_ia1, &self.expr_ia2] {
+            let res = child.eval_checked(&input)?;
+            let orig_vis = input.vis();
+            let res_vis: Vis = match self.expr_type {
+                // If res of left part is not null and is true, we do not want to calculate right
+                // part because the result must be true.
+                Type::Or => (!(res.as_bool().to_bitmap())).into(),
+                _ => unimplemented!(),
+            };
+            let new_vis = orig_vis & res_vis;
+            input.set_vis(new_vis);
+            children_array.push(res);
+        }
+        let mut builder =
+            <BoolArray as Array>::Builder::with_meta(len, (&self.return_type()).into());
+        match self.expr_type {
+            Type::Or => {
+                for (((v_ia1, v_ia2), init_visible), final_visible) in multizip((
+                    children_array[0].as_bool().iter(),
+                    children_array[1].as_bool().iter(),
+                ))
+                .zip_eq(init_vis.iter())
+                .zip_eq(input.vis().iter())
+                {
+                    if init_visible {
+                        builder.append(if final_visible {
+                            or(v_ia1, v_ia2)?
+                        } else {
+                            Some(true)
+                        });
+                    } else {
+                        builder.append_null()
+                    }
+                }
+            }
+            _ => unimplemented!(),
+        }
+        Ok(Arc::new(builder.finish().into()))
+    }
+
+    fn eval_row(&self, input: &Row) -> Result<Datum> {
+        let ret_ia1 = self.expr_ia1.eval_row(input)?.map(|x| x.into_bool());
+        match self.expr_type {
+            Type::Or => {
```

Comment:
Can we use `assert` instead of `match`? I think `assert` can make the code clearer.

Reply:
> Can we use `assert` instead of `match`? I think `assert` can make the code clearer.

`Type::And` will be added here in the foreseeable future

Reply:
Cool.